src/main/java/org/folio/services/financedata/FinanceDataValidator.java

```diff
@@ -0,0 +1,181 @@
+package org.folio.services.financedata;
+
+import static io.vertx.core.Future.failedFuture;
+import static io.vertx.core.Future.succeededFuture;
+import static org.folio.rest.util.ErrorCodes.BUDGET_STATUS_INCORRECT;
+import static org.folio.rest.util.HelperUtils.collectResultsOnSuccess;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import io.vertx.core.Future;
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.folio.rest.core.models.RequestContext;
+import org.folio.rest.exception.HttpException;
+import org.folio.rest.jaxrs.model.Error;
+import org.folio.rest.jaxrs.model.Errors;
+import org.folio.rest.jaxrs.model.FyFinanceData;
+import org.folio.rest.jaxrs.model.FyFinanceDataCollection;
+import org.folio.rest.jaxrs.model.Parameter;
+import org.folio.rest.jaxrs.model.SharedBudget;
+import org.folio.services.budget.BudgetService;
+import org.folio.services.fund.FundService;
+
+@Log4j2
+public class FinanceDataValidator {
+
+  private final FundService fundService;
+  private final BudgetService budgetService;
+
+  public FinanceDataValidator(FundService fundService, BudgetService budgetService) {
+    this.fundService = fundService;
+    this.budgetService = budgetService;
+  }
+
+  public void validateFinanceDataCollection(FyFinanceDataCollection financeDataCollection, String fiscalYearId) {
+    validateForDuplication(financeDataCollection);
+    for (int i = 0; i < financeDataCollection.getFyFinanceData().size(); i++) {
+      var financeData = financeDataCollection.getFyFinanceData().get(i);
+      validateFinanceDataFields(financeData, i, fiscalYearId);
+
+      if (financeData.getBudgetAllocationChange() != null) {
+        var allocationChange = financeData.getBudgetAllocationChange();
+        var currentAllocation = financeData.getBudgetCurrentAllocation();
+
+        if (allocationChange < 0 && Math.abs(allocationChange) > currentAllocation) {
+          var error = createError("Allocation change cannot be greater than current allocation",
+            String.format("financeData[%s].budgetAllocationChange", i), String.valueOf(financeData.getBudgetAllocationChange()));
+          throw new HttpException(422, new Errors().withErrors(List.of(error)));
+        }
+      }
+    }
+  }
+
+  private void validateForDuplication(FyFinanceDataCollection financeDataCollection) {
+    var financeFundBudgetFiscalYearIds = financeDataCollection.getFyFinanceData()
+      .stream().collect(Collectors.groupingBy(
+        financeData -> financeData.getFundId() + financeData.getBudgetId() + financeData.getFiscalYearId()));
```

Comment:
would null for budgetId be considered as duplicates? we should not consider nulls as duplicated because have this story: https://folio-org.atlassian.net/browse/MODFISTO-517